File: 0136/hash_table.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int singleNumber(vector<int>& nums) {
    unordered_map<int, int> map;
    for (const auto& num : nums) map[num]++;
    int ret = 0;
    for (const auto& pair : map) {
      if (pair.second == 1) ret = pair.first;
    }
    return ret;
  }
};

int main() {
  Solution sol;
  vector<int> nums;

  nums = {2,2,1};
  assert(sol.singleNumber(nums) == 1);

  nums = {4,1,2,1,2};
  assert(sol.singleNumber(nums) == 4);

  nums = {2};
  assert(sol.singleNumber(nums) == 2);
}

File: 0260/hash_table.py
from collections import defaultdict
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        counts = defaultdict(int)
        for num in nums:
            counts[num] += 1
        ret = []
        for num, count in counts.items():
            if count == 1:
                ret.append(num)
        return ret


if __name__ == '__main__':
    sol = Solution()

    nums = [1, 2, 1, 3, 2, 5]
    expected = [3, 5]
    assert sol.singleNumber(nums) == expected
